File: python/rss_utils.py
# File: utils/rss_utils.py

import requests
import time
import xml.etree.ElementTree as ET
from urllib.parse import quote_plus, unquote

def validate_rss_feed(url):
    """
    Validate RSS feed by checking both accessibility and content.
    Returns True only if the feed is accessible and contains torrent items.
    """
    try:
        response = requests.get(url, timeout=5)
        if response.status_code != 200 or 'xml' not in response.headers.get('content-type', ''):
            return False, None

        # Parse XML content
        root = ET.fromstring(response.content)
        
        # Find channel and then items
        channel = root.find('channel')
        if channel is None:
            return False, None
            
        # Look for item elements
        items = channel.findall('item')
        episode_count = len(items)
        
        # Return True and episode count only if there are items with valid links
        if episode_count > 0 and any(item.find('link') is not None for item in items):
            return True, episode_count
        return False, None

    except (requests.RequestException, ET.ParseError):
        return False, None
    
def extract_from_rss_url(rss_url):
    """Extract search terms and fansubber from RSS URL."""
    try:
        # Get the query part after q= and before &
        q_param = rss_url.split('q=')[1].split('&')[0]
        # URL decode the parameter
        q_param = unquote(q_param)
        # Split on spaces
        parts = q_param.split('+')
        
        # Find the index after -batch
        batch_index = parts.index('-batch')
        if batch_index + 1 >= len(parts):
            return None, None
            
        # Get fansubber (first term after -batch)
        fansubber = parts[batch_index + 1]
        
        # Get search terms (everything after fansubber)
        search_terms = '+'.join(parts[batch_index + 2:])
        
        return fansubber, search_terms
    except:
        return None, None
    
def initialize_fansubber_from_rss(rss_url, fansubber_key):
    """Extract and initialize fansubber value from RSS URL."""
    fansubber, _ = extract_from_rss_url(rss_url)
    if fansubber:
        return fansubber
    return "ember"  # Default fallback

def validate_rss_feeds_with_delay(rss_urls, delay=0.5):
    """Validate RSS feeds with a delay between each request."""
    results = {}
    episode_counts = {}
    total = len(rss_urls)
    
    for i, url in enumerate(rss_urls, 1):
        is_valid, count = validate_rss_feed(url)
        results[url] = is_valid
        episode_counts[url] = count
        time.sleep(delay)
        yield results, episode_counts, (i / total) * 100

def format_rss_url(title, fansubber="ember"):
    """Format the RSS URL for a given anime title and fansubber."""
    # For RSS URLs we need to properly encode everything including spaces
    safe_fansubber = quote_plus(fansubber)
    # Replace quotes first, then encode the whole title
    safe_title = title.replace('"', "'")
    safe_title = quote_plus(safe_title)

    return f"https://nyaa.si/?page=rss&q=-batch+{safe_fansubber}+{safe_title}&c=0_0&f=0"

def format_search_url(title, fansubber="ember"):
    """Format the search URL for a given anime title and fansubber."""
    # Encode the fansubber and title separately
    safe_fansubber = quote_plus(fansubber)
    # For the title, we want to keep spaces for the search interface
    safe_title = title.replace('"', "'")  # Replace quotes but keep spaces
    
    # Construct the URL keeping spaces between terms
    return f"https://nyaa.si/?f=0&c=0_0&q=-batch+{safe_fansubber}+{safe_title}"

def format_search_url_from_terms(search_terms, fansubber):
    """Format search URL from extracted search terms and fansubber."""
    # URL encode fansubber but keep spaces in search terms
    safe_fansubber = quote_plus(fansubber)
    # For search terms, replace quotes but keep spaces
    safe_terms = search_terms.replace('"', "'")
    
    return f"https://nyaa.si/?f=0&c=0_0&q=-batch+{safe_fansubber}+{safe_terms}"

def validate_rss_url(url):
    """Validate if the URL contains 'rss' and basic URL structure."""
    return 'rss' in url.lower() and (url.startswith('http://') or url.startswith('https://'))

File: lib/services/rss_utils.dart
import 'package:http/http.dart' as http;
import 'package:xml/xml.dart' as xml;

class RssUtils {
  static Future<(bool, int?)> validateRssFeed(String url) async {
    try {
      final response = await http
          .get(Uri.parse(url))
          .timeout(const Duration(seconds: 5));

      if (response.statusCode != 200 ||
          !response.headers['content-type']!.contains('xml')) {
        return (false, null);
      }

      // Parse XML content
      final document = xml.XmlDocument.parse(response.body);

      // Find channel and then items
      final channelElements = document.findAllElements('channel');
      if (channelElements.isEmpty) {
        return (false, null);
      }

      final channel = channelElements.first;

      // Look for item elements
      final items = channel.findAllElements('item').toList();
      final episodeCount = items.length;

      // Return true and episode count only if there are items with valid links
      if (episodeCount > 0 &&
          items.any((item) => item.findElements('link').isNotEmpty)) {
        return (true, episodeCount);
      }

      return (false, null);
    } catch (e) {
      return (false, null);
    }
  }

  static (String?, String?) extractFromRssUrl(String rssUrl) {
    try {
      // Get the query part after q= and before &
      final qParamStart = rssUrl.indexOf('q=');
      if (qParamStart == -1) return (null, null);

      final qParamEnd = rssUrl.indexOf('&', qParamStart);
      final qParam = Uri.decodeComponent(
        qParamEnd == -1
            ? rssUrl.substring(qParamStart + 2)
            : rssUrl.substring(qParamStart + 2, qParamEnd),
      );

      // Split on spaces
      final parts = qParam.split('+');

      // Find the index after -batch
      final batchIndex = parts.indexOf('-batch');
      if (batchIndex == -1 || batchIndex + 1 >= parts.length) {
        return (null, null);
      }

      // Get fansubber (first term after -batch)
      final fansubber = parts[batchIndex + 1];

      // Get search terms (everything after fansubber)
      final searchTerms = parts.sublist(batchIndex + 2).join('+');

      return (fansubber, searchTerms);
    } catch (e) {
      return (null, null);
    }
  }

  static String initializeFansubberFromRss(String rssUrl) {
    final (fansubber, _) = extractFromRssUrl(rssUrl);
    return fansubber ?? 'ember';
  }

  static String formatRssUrl(String title, [String fansubber = 'ember']) {
    // For RSS URLs we need to properly encode everything including spaces
    final safeFansubber = Uri.encodeComponent(fansubber);
    // Replace quotes first, then encode the whole title
    final safeTitle = Uri.encodeComponent(title.replaceAll('"', "'"));

    // New URL format for Animetosho
    return 'https://feed.animetosho.org/rss2?only_tor=1&q=-batch+$safeFansubber+$safeTitle';
  }

  static String formatSearchUrl(String title, [String fansubber = 'ember']) {
    // Encode the fansubber
    final safeFansubber = Uri.encodeComponent(fansubber);
    // For the title, we want to keep spaces for the search interface
    final safeTitle = title.replaceAll('"', "'");

    // Construct the search URL for Animetosho web interface
    return 'https://animetosho.org/search?q=-batch+$safeFansubber+$safeTitle';
  }

  static String formatSearchUrlFromTerms(String searchTerms, String fansubber) {
    // URL encode fansubber
    final safeFansubber = Uri.encodeComponent(fansubber);
    // For search terms, replace quotes but keep spaces
    final safeTerms = searchTerms.replaceAll('"', "'");

    return 'https://animetosho.org/search?q=-batch+$safeFansubber+$safeTerms';
  }

  static bool validateRssUrl(String url) {
    return url.toLowerCase().contains('rss') &&
        (url.startsWith('http://') || url.startsWith('https://'));
  }

  static String convertRssToSearchUrl(String rssUrl) {
    try {
      // Check if it's an Animetosho RSS URL
      if (rssUrl.contains('feed.animetosho.org/rss2')) {
        // Parse the RSS URL
        final uri = Uri.parse(rssUrl);

        // Get the query parameters
        final queryParams = uri.queryParameters;

        // We need the search query from the q parameter
        final searchQuery = queryParams['q'] ?? '';

        // Construct the search URL
        return 'https://animetosho.org/search?q=$searchQuery';
      } else {
        // Fallback for other RSS sources
        return rssUrl.replaceFirst('rss2', 'search');
      }
    } catch (e) {
      print('Error converting RSS to search URL: $e');
      return rssUrl; // Return original URL on error
    }
  }
}
